make rule deprecated-local-action preserve module parameters

koalajoe23 opened the pull request:
Proposed fix for #4731

This is my first attempt at hacking something I want into ansible-lint, be gentle ;)

I was a bit surprised why all rule tests get a a full ruleset provided via `default_rules_collection` param as this makes the rules prone to breakage caused by other rules' auto-formating. I changed this and also made some assertions test if `decprecated-local-action` actually matched, not the number of expected matches which may change in the future.

koalajoe23 commented:
I think the remaining CI errors are not cause by my PR, right?

koalajoe23 commented:
I changed the debug log statements for missing "local_action" to assertions.

Code coverage went down because of the log statements and while trying to write test fixtures, I realized a missing `local_action` would result in a non-matching rule, and a missing `module` param would be caught by the `parser-error` rule, also never reaching `deprecated-local-action`.

I think the assertions are fine here to guard against implementation errors.

cidrblock commented:
I pulled it down to add some examples and realized this is much more complicated than I roginally thought.

I updated your PR with some edge cases, and got rid of the runtime "assert" statements and added some early exits whic tend to ease the type hinting.

Hope it's ok, let me know what you think.  The only issue now is because I committed I need a peer review from someone else on the team :)

koalajoe23 commented:
Hi,

> I updated your PR with some edge cases

:+1: 

I appreciate! didn't think of legacy "dict-as-string" code and throwing other dict elements in and I think it improves the transform test robustness

> got rid of the runtime "assert" statements and added some early exits which tend to ease the type hinting

:+1: 

My intention was to have both cases fail rather loudly, as both assertions should never trigger, if not caused by an implementation error outside of this rule's `transform()`, not by user input. I was a bit misguided by how assertions are used in C++ and now agree that using them outside of testing code or code ran with optimizations enabled is not a good idea in Python.

[nitpick]: Maybe a log statement could be added to

```python
if "local_action" not in original_task:
  return
```
> Hope it's ok, let me know what you think.

In my understanding (please correct me if wrong), the linter rule transform should do one thing only: transform the ansible task matched by that rule, so that it no longer matches without introducing side-effects or altering the outcome, while preserving everything else. That transformed task may be matched by another rule afterwards, but a single rule should focus on it's core responsibility.

The code you proposed does too much, imo:
1. **[critical]** Your introduction of `shlex` removed parentheses from string-form arguments. This could introduce bugs and security vulnerabilities into the users code! `shell: echo "Hello World; /usr/bin/self-destruct"` is not the same as `shell: echo Hello World; /usr/bin/self-destruct`
2. You move the `name` param to the top, which should be the responsibility of rule `key-order` - which might be explicitly disabled by the user.
3. You make `dict`s out of the "key=val" free-form parameter string. Should be handled by `no-free-form` rule.
4.  [nitpick]: I think the `delegate_to` param should be inserted right after the new module param to make diffs a bit easier to parse.

**Only point no 1 is a merge-blocker for me.** I'd be glad to get feedback on my other remarks though! In the end it's on you as maintainer(s) to decide. Functionality-wise, the rule would be fine for me as-is, as it won't break the ansible playbooks I inherit from  other people from time to time again :)

cidrblock commented:
> [nitpick]: Maybe a log statement could be added to
> 
> ```python
> if "local_action" not in original_task:
>   return
> ```
> 

I don't know the code well enough, but we should never be in the transform for local_action if there isn't one, I can probable rm that.

> In my understanding (please correct me if wrong), the linter rule transform should do one thing only: transform the ansible task matched by that rule, so that it no longer matches without introducing side-effects or altering the outcome, while preserving everything else. That transformed task may be matched by another rule afterwards, but a single rule should focus on it's core responsibility.

Completely correct.
> 
> The code you proposed does too much, imo:

Agree.
> 
> 1. **[critical]** Your introduction of `shlex` removed parentheses from string-form arguments. This could introduce bugs and security vulnerabilities into the users code! `shell: echo "Hello World; /usr/bin/self-destruct"` is not the same as `shell: echo Hello World; /usr/bin/self-destruct`

Good catch!

> 2. You move the `name` param to the top, which should be the responsibility of rule `key-order` - which might be explicitly disabled by the user.

Right again.

> 3. You make `dict`s out of the "key=val" free-form parameter string. Should be handled by `no-free-form` rule.

and again.

> 4. [nitpick]: I think the `delegate_to` param should be inserted right after the new module param to make diffs a bit easier to parse.

Fail.
> 
> **Only point no 1 is a merge-blocker for me.** I'd be glad to get feedback on my other remarks though! In the end it's on you as maintainer(s) to decide. Functionality-wise, the rule would be fine for me as-is, as it won't break the ansible playbooks I inherit from other people from time to time again :)

rework coming.  Thanks for the detailed review.

cidrblock commented:
@koalajoe23 Does this do everything you need now?  Fun stuff.  BTW, your intuition was right about needing a way to get out quick, it part of the reason I threw in an exception, it also made adding some unit tests for the edge cases pretty straight forward with having to traverse the entire code base.

koalajoe23 commented:
:+1: for the changes. I really like the decoupled tests and error handling!

Only nitpick I have is that the transformed output should treat empty params as None - just looks much cleaner and "human". I'll take the liberty to just commit a fix, feel free to drop this commit if you want.

Everything LGTM on my part, awaiting merge :)
